cellmlmanip/units.py

```diff
@@ -622,3 +635,179 @@ def traverse(self, expr):
+        dimensionless = self._store.get_unit('dimensionless')
+
+        def maybe_convert_expr(expr, was_converted, to_units, from_units):
+            try:
+                cf = self._store.get_conversion_factor(to_units, from_units)
+            except DimensionalityError:
+                raise UnitConversionError(expr, from_units, to_units) from None
+            if cf != 1.0:
+                print('Converting {} to {} by factor {}'.format(from_units, to_units, cf))
+                was_converted = True
+                cf = model.NumberDummy(cf, to_units / from_units)
+                expr = cf * expr
+            return expr, was_converted, to_units
+
+        def maybe_convert_child(expr, was_converted, to_units):
+            expr, child_was_converted, actual_units = self._convert_expr_recursively(expr, to_units)
+            return expr, child_was_converted or was_converted, actual_units
+
+        if expr.is_Matrix:
+            # See comment in :meth:`traverse()`!
+            raise UnexpectedMathUnitsError(str(expr))
+        elif expr.is_Symbol:
+            assert isinstance(expr, (model.NumberDummy, model.VariableDummy))
+            if to_units is None:
+                # Nothing to do, just record the actual units
+                actual_units = expr.units
+            else:
+                expr, was_converted, actual_units = maybe_convert_expr(expr, was_converted, to_units, expr.units)
+        elif expr.is_Derivative:
+            # Just convert the result if needed
+            _, was_converted, numerator_units = maybe_convert_child(expr.args[0], was_converted, None)
+            _, was_converted, denominator_units = maybe_convert_child(expr.args[1][0], was_converted, None)
+            assert not was_converted  # Ensures both parts are simple symbols
+            actual_units = numerator_units / denominator_units
+            if to_units is not None:
+                expr, was_converted, actual_units = maybe_convert_expr(expr, was_converted, to_units, actual_units)
+        elif expr.is_Mul:
+            # Multiply units of operands, which are allowed to be of any units
+            all_arg_units = []
+            for arg in expr.args:
+                arg, was_converted, arg_units = maybe_convert_child(arg, was_converted, None)
+                new_args.append(arg)
+                all_arg_units.append(arg_units)
+            actual_units = reduce(mul, all_arg_units)
+            if to_units is not None or was_converted:
+                # Convert the result if required
+                expr = expr.func(*new_args)
+                expr, was_converted, actual_units = maybe_convert_expr(expr, was_converted, to_units, actual_units)
+        elif expr.is_Pow:
+            # Pow is used by Sympy for exponentiating, roots and division
+            base, exponent = expr.args
+            # Exponent must be dimensionless
+            exponent, was_converted, _ = maybe_convert_child(exponent, False, dimensionless)
```

Comment:
Should we be catching the possible errors in cases like this, and raising more useful ones?
Now I guess it'll say something like "can't convert 1mV to dimensionless", and then the user has to guess why that conversion was required at some point?

Reply:
Seems like we could easily have something like "Power requires dimensionless exponent, but got ..."

Reply:
All the conversion errors do report the (sub-)expression they're trying to convert. If you also had the full expression as mentioned above, would this be sufficient context for this PR? We could add more in a future PR if it turned out to be needed in actual usage?

Reply:
Have added more informative messages now; see what you think.